### DISCUSSION TITLE
Dump SNES Repro : message "Does not have ROM" & "Banks: 0" (RESOLVED)

### DISCUSSION THREAD

**Discussion by StanMOI** [sanni/cartreader#164] [Need help]
Hi Sanni,
Here again I have an error message "Does not have ROM" when I want to read SNES Repro.
![IMG_6222](https://user-images.githubusercontent.com/38383423/120338985-03b67f80-c2f5-11eb-8d9f-7561fa38798a.jpeg)

Repro PCB have game "Fire Emblem" with ROM size 64Mbit (works fine on TV)

Your reader see name of game but say that "Banks: 0" 
Flash eprom 29LV640
![IMG_6225](https://user-images.githubusercontent.com/38383423/120338624-b33f2200-c2f4-11eb-854c-c8fe047ee29a.jpeg)
![IMG_6226](https://user-images.githubusercontent.com/38383423/120338638-b76b3f80-c2f4-11eb-9e84-c74407b5f273.jpeg)
![IMG_6223](https://user-images.githubusercontent.com/38383423/120338957-fc8f7180-c2f4-11eb-93c3-06b0399ad304.jpeg)

For compare I tried to read an other game.
I see that it can detect all banks "Banks: 32" 
![IMG_6224](https://user-images.githubusercontent.com/38383423/120339294-46785780-c2f5-11eb-987a-f3a046dcd809.jpeg)

Whats wrong ?

When I erase an **other identical PCB** with same Flash EPROM ans SRAM (64Mbit with 256 SRAM),
And go to "Super Nintendo/Super Nintendo/"
I have this error message :
![IMG_6228](https://user-images.githubusercontent.com/38383423/120340498-60ff0080-c2f6-11eb-9f38-739868e36ecf.jpg)
Then I press button and get this one, with also "Banks: 0":
![IMG_6227](https://user-images.githubusercontent.com/38383423/120340574-73793a00-c2f6-11eb-99cf-ab0ed75a9b2c.jpg)
I dont know if it's normal message for erased flash?

After what I go to "write eprom" with identical size file 6,3Mo (64Mbit) "Fire Emblem".
Programming end test at the end works fine.
When I try to read this PCB , its recognize game on flash 
![IMG_6223](https://user-images.githubusercontent.com/38383423/120338957-fc8f7180-c2f4-11eb-93c3-06b0399ad304.jpeg)
And get same error "Does not have ROM" with "Banks: 0" and nothing on TV, only black screen.
![IMG_6222](https://user-images.githubusercontent.com/38383423/120338985-03b67f80-c2f5-11eb-8d9f-7561fa38798a.jpeg)

Thank you for help

**Comment by sanni** [comment]
You can add the number of banks and the size to [snes.txt](https://github.com/sanni/cartreader/blob/master/sd/snes.txt)

Example:
A172,24,48
checksum, size in Mbit, number of banks (lorom needs twice as many banks as hirom for the same rom size, lorom: 32kb banks, hirom: 64kb banks)

The first hex value would be E390 in your case, the next the size of your rom file times 8, so if it's a 4MB rom file it would be 32, and lastly the numer of banks for lorom it would be rom size in MB times 1024 divided by 32.

**Reply by StanMOI** [reply]
> You can add the number of banks and the size to [snes.txt](https://github.com/sanni/cartreader/blob/master/sd/snes.txt)
> 
> Example:
> A172,24,48
> checksum, size in Mbit, number of banks (lorom needs twice as many banks as hirom for the same rom size, lorom: 32kb banks, hirom: 64kb banks)
> 
> The first hex value would be E390 in your case, the next the size of your rom file times 8, so if it's a 4MB rom file it would be 32, and lastly the numer of banks for lorom it would be rom size in MB times 1024 divided by 32.

E390,64

The last I dont understand ...
_"it would be rom size in MB"_ = 64 (times 1024)
_"divided by 32"_ = 64Mbit/32=2

rom size in MB times 1024 ?

Is that it ?
E390,64,256 (Checksum, Size in Mbit, Lorom 64Mbit ROM)

**Comment by sanni** [comment]
Yes E390,64,256 would be correct but I don't know if the current code can handle so many banks. ExLoRom is not supported and never has been tested. Which is probably why it shows 0 banks, highest bank count supported right now is 128, which results in a 4MB game with LoRom.

**Reply by StanMOI** [reply]
> Yes E390,64,256 would be correct but I don't know if the current code can handle so many banks. ExLoRom is not supported and never has been tested. Which is probably why it shows 0 banks, highest bank count supported right now is 128, which results in a 4MB game with LoRom.

I will try and give you feedback

**Comment by StanMOI** [comment]
This code in .txt file are only for "read" or for writing too ?

**Comment by StanMOI** [comment]
Here is results after changing **snes.txt** file with adding number of banks **E390,64,256** :

- Its still shows "0 banks" (when analyzing chip).

- When I try to program chip (by Writing command) and test it after on SNES = still black screen on TV
- After that I try to analyse programmed PCB again with "Super Nintendo/Super Nintendo/", shows "0 banks" .

oh no....

PS: can you add fonction for ExLoRom like you done with exHiROM, please?

**Comment by StanMOI** [comment]
I tried to write on this PCB rom with **16Mbit** and 32Mbit (LoROM with SRAM 256), the result is = black screen 
Dose it mean that 29LV640 are not supported ?
Or I have to add some new lines to **snes.txt** file, like:
E390,32,256
E390,32,128

**Comment by sanni** [comment]
In [Cart_Reader.ino](https://github.com/sanni/cartreader/blob/master/Cart_Reader/Cart_Reader.ino#L212) line 212 change "byte numBanks = 128;" to "word numBanks = 128;".
Then try dumping again to see if the banks are displayed correctly now.
It's normal that deleted cartridges don't show any info, you have to try with a PCB that still has the game flashed.

snes.txt only works for dumping SNES. The Repro menu uses completely separate code. You can't flash LoRom games to an ExLoRom PCB unless you extent them to 8MB since LoRom games don't address A22 rom/A23 snes or unless you change a jumper on the PCB to keep A22 rom/A23 snes high.
If flashing an 8MB rom file does not give you an error while flashing then flashing exLoRom should work fine.

I can't help you in more detail because I don't have any exLoRom games/PCBs.

**Reply by StanMOI** [reply]
> In [Cart_Reader.ino](https://github.com/sanni/cartreader/blob/master/Cart_Reader/Cart_Reader.ino#L212) line 212 change "byte numBanks = 128;" to "word numBanks = 128;".
> Then try dumping again to see if the banks are displayed correctly now.
> It's normal that deleted cartridges don't show any info, you have to try with a PCB that still has the game flashed.
> 
> snes.txt only works for dumping SNES. The Repro menu uses completely separate code. You can't flash LoRom games to an ExLoRom PCB unless you extent them to 8MB since LoRom games don't address A22 rom/A23 snes or unless you change a jumper on the PCB to keep A22 rom/A23 snes high.
> If flashing an 8MB rom file does not give you an error while flashing then flashing exLoRom should work fine.
> 
> I can't help you in more detail because I don't have any exLoRom games/PCBs.

Thank you Sanni for help,
I changed to "word numBanks = 128;"
Now its showing "Banks: 256" and I can dump it, its works fine!
After what I tied to flash with 6.3Mo rom : Still black screen on TV, but recognize when analyzing PCB.

So if I understand right, please correct me if not :
I can not flash exLoROM rom's (with 6.3Mo), cause your **Cart Reader** not support it.
To flash **simple LoROM rom** to exLoROM PCB i have two solutions : 
- extend file to 8MB (64Mbit), but flashing are not supported by **Cart Reader**
- Cut **jumper 22** 
 ![IMG_6235](https://user-images.githubusercontent.com/38383423/120457433-bb4f9e00-c396-11eb-8995-e7dee41d7b34.jpg)
Jumpers Connections : 
jumper 22 : CART 48 => Flash pin 13 (A22)
jumper 21  (dont need to be touch): CART 47 => Flash pin 10 (A21)
jumper 20 (dont need to be touch) : CART 46 => Flash pin 9 (A20)


PS: 
I already tried to do it by Cutting Jumper 22 (CART 48 => Flash pin 13).
After what, when I try to Flash it with 16Mbit (2,1Mo) or 32Mbit (4,2Mo ) LoROM rom,
its freezing during flashing (something like at the end).
No led blinking.
![IMG_6238](https://user-images.githubusercontent.com/38383423/120468974-15099580-c3a2-11eb-8876-b4fad3ff8b98.jpg)
![IMG_6237](https://user-images.githubusercontent.com/38383423/120468995-1b980d00-c3a2-11eb-9044-008a7a296794.jpg)

Maybe I should do like that : 
First => Flash rom with soldered Jumper 22 (cause Flashing works like that: no freezing, step of verifying at the end work's)
Second (after that it was flashed)  => desolder jumper 22.

PS bis:
Already test it, still have Black screen on TV
When I flash 16Mbit rom with connected jumper 22  (flashing works with no error),
Then desolder this jumper and test it with "Super Nintendo/Super Nintendo/", I get this message:
![IMG_6239](https://user-images.githubusercontent.com/38383423/120471990-b0e8d080-c3a5-11eb-8d03-0386d2a8142f.jpg)
![IMG_6240](https://user-images.githubusercontent.com/38383423/120472369-1b017580-c3a6-11eb-97f2-9a98833914af.jpg)

**Comment by sanni** [comment]
You can try the following:
- put solder jumper back
- increase size of 2MB LoRom to 8MB, for example by opening it in HxD Hexeditor and CTRL+A, CTRL+C and then multiple CTRL+V until size of 8MB is reached.
- then flash and see what happens

If it works you can try the following:
- increase size of 6MB ExLoRom to 8MB by adding 0xFF at the end until exactly 8MB is reached, split the rom in two 4MB parts, join the two parts in reverse order together to 8MB
- then flash and see what happens

**Reply by StanMOI** [reply]
> You can try the following:
> 
> * put solder jumper back
> * increase size of 2MB LoRom to 8MB, for example by opening it in HxD Hexeditor and CTRL+A, CTRL+C and then multiple CTRL+V until size of 8MB is reached.
> * then flash and see what happens

Done,
Just increase 2MB LoRom to 8MB = nothing changes, black screen.
I tried to desolder jumper 22 again (after flashing), the result is the same = black screen.

> * adding 0xFF at the end 

Dont know how to do it, here is my two Hex Editors (for Mac).
I can not write "x"

<img width="262" alt="Capture d’écran 2021-06-02 à 16 08 06" src="https://user-images.githubusercontent.com/38383423/120501989-5742cf00-c3c2-11eb-92ef-7e75c448a572.png">
<img width="262" alt="Capture d’écran 2021-06-02 à 16 08 52" src="https://user-images.githubusercontent.com/38383423/120502006-59a52900-c3c2-11eb-9c60-88739128e7cd.png">
<img width="531" alt="Capture d’écran 2021-06-02 à 16 46 58" src="https://user-images.githubusercontent.com/38383423/120502031-5f9b0a00-c3c2-11eb-86cb-15880b6574f5.png">

I used "Lunar Expand" software to expand file 6MB ExLoRom to 8MB (in version 1 of exLoROM maps, dont know if I should try with other version of maps) .
Once expanded, I check it in Hex Editor and see only 0000  at the end of file (Dont know if its normal).
![IMG_6248](https://user-images.githubusercontent.com/38383423/120506932-aab71c00-c3c6-11eb-8dd5-8b25deadc163.jpg)

Then tried to flash expanded file (8MB version 1) With **jumper 22** connected.
Then I've done what you said : splited file in two 4MB parts, than joined two parts in reverse order together to 8MB.

Here is results of testing (with jumper 22 connected):
**Lunar Expand 8MB Vesion 1** 
Flash = ok (no bug, no freezing, verifying ok)
Recognize after flashing by Cart Reader = ok (no error, Banks: 256)
TV test = Black screen

PS :
i have the similar PCB LoROM with 32Mbit flash MX29LV320, (not exLoROM) with **jumper 22 disconnected**.
Other connections vs exLoROM PCB are the same (except this jumper 22).
Flashing works fine. TV test works fine. Recognizing by Cart Reader and Dumping works fine.
I can Flash this PCB (MX29LV320) with 16Mbit or 32Mbit rom.
Why I can not flash exLoROM PCB with an simple rom (Lorom) if the only difference are between PCB's is chip size MX29LV640 (exLoROM PCB) vs MX29LV320 (LoROM PCB),
and this « jumper 22 » with connection **CART 48** => **Flash pin 13 (A22**) when its exLoROM.

**Reply by sanni** [reply]
LV640 has one address line more than LV320. If you just desolder the jumper this address is floating, meaning it toggles between High and Low all the time. You need to connect the top pad of the jumper either to GND or to VCC to give it a stable value. But it's important that the solder jumper is removed so that the bottom pad is not connected to anything.

**Reply by StanMOI** [reply]
> LV640 has one address line more than LV320. 

LV320 (and LV160), pin 13 = NC (not connected)
LV640, pin 13 = A22

> If you just desolder the jumper this address is floating, meaning it toggles between High and Low all the time. You need to connect the top pad of the jumper either to GND or to VCC to give it a stable value. 

Please correct me if Its wrong,
You mean that for PCB with **LV640**, when I use it like simple LoROM (8, 16, 32Mbit), 
I have to desolder "jumper 22" and connect the **top pad** of this jumper to VCC or GND.

> important that the solder jumper is removed so that the bottom pad is not connected to anything.

CART 48 = NC (when I using  **LV640** like simple LoROM : 8, 16, 32Mbit)

**Comment by sanni** [comment]
Yes, by doing that the 8MB flashrom becomes a 4MB flashrom. If the rest of the PCB is the same then it should work.
You can also connect a switch that switches the top pad between GND and VCC, then you have basically two 4MB roms and can flash two games to it and select which one you want to play by switching.

**Reply by StanMOI** [reply]
> Yes, by doing that the 8MB flashrom becomes a 4MB flashrom. If the rest of the PCB is the same then it should work.
> You can also connect a switch that switches the top pad between GND and VCC, then you have basically two 4MB roms and can flash two games to it and select which one you want to play by switching.

Thank you Sanni,
I'll do it tomorrow and give you feedback.

Sad that exLoROM are not supported...
Maybe once you will have a time to do new version.
So you can add all that you told me for repros (like adding new lines to snes.txt file, etc)

**Comment by StanMOI** [comment]
Do i have to expand files every time ?
I mean, if fox example I have 8Mbit or 16Mbit rom and want flash it into LV320 (32Mbit), 
Do I have to increase file size to exactly flash size? 

I thought that Flash Eprom doesn't works like UV EPROM's.
I thought that using Flash Eprom expanding wasn't necessary.

**Comment by sanni** [comment]
Expanding was just a test ,you don't need to do that normally.
Adding new lines to SNES.txt was also just a test, the real bug was the "byte" that was replaced by "word" to make reading working.

There is one more test you can do. Write a 8MB rom file to the ExLoRom cartridge, then dump it and compare the dump to the rom file using an hex editor like HxD and the Compare function. Just to make sure writing ExLoRom actually works even if it doesn't boot.

Also when the repro has a black screen in the SNES be sure to hold the reset button in case the snesCIC is switched to the wrong region.

**Reply by StanMOI** [reply]
> Expanding was just a test ,you don't need to do that normally.
> Adding new lines to SNES.txt was also just a test, the real bug was the "byte" that was replaced by "word" to make reading working.
> 
> There is one more test you can do. Write a 8MB rom file to the ExLoRom cartridge, then dump it and compare the dump to the rom file using an hex editor like HxD and the Compare function. Just to make sure writing ExLoRom actually works even if it doesn't boot.
> 
> Also when the repro has a black screen in the SNES be sure to hold the reset button in case the snesCIC is switched to the wrong region.

Thanks Sanni,

Exactly, about exLoROM that was it !
I tested dumped file = it was broken, game not starts (should be me, mistakes some where).
So I'he tried again : 
extend rom to 8Mbit => than split to 4MB (so I have 2x4MB) => than inverted files and joined together to get 8MB file => Than Flash my exLoROM board => then tried to Dump board again and see if file is OK (dump=ok, YES file works!) => than TV test...and it's WORKS !!!!
Hallelujah !!!

After what I tried next step: 
Flash exLoROM (LV640) board with 16Mbit (2MB) LoROM rom (without expanding that).

Test 1:
Jumper 22 (cut) : pin 13 of Flash = VCC, CART 48 = NC
Doing that 8MB flashrom becomes a 4MB flashrom.
In that configuration;
Erase= works
Flashing= ok
Verifying = ok
TV test = its works!
DUMP : error, but after next screen it's recognize rom

Test 2:
Jumper 22 (solder) : pin 13 of Flash = CART 48
Doing that 8MB flashrom stay 8MB flashrom.
In that configuration;
Erase= works
Flashing= ok
Verifying = ok
TV test = its works!
DUMP : error, but after next screen it's recognize rom

Surprise that in Test 2, CART 48 are connected to pin 13 and its works (most be something aleatory).
After what I changed **Jumper 22 (cut)** : pin 13 of Flash = VCC, CART 48 = NC
TV test = black screen
DUMP : error, unknown rom

Haha

**Comment by sanni** [comment]
So in your opinion should I list ExLoRom as supported now or keep it as unsupported?

**Reply by StanMOI** [reply]
> So in your opinion should I list ExLoRom as supported now or keep it as unsupported?

I think YES, but I can do more tests if you need it.

Sanni,
Its will be great if you add Write ExLoRom option to your reader (like exHiROM).
Will you do it in future ?

**Comment by StanMOI** [comment]
Sanni,
Why when I flashed board LV640 with 16Mbit (2MB) LoROM rom. (flashing works with no error, TV test=OK)
It's impossible to Dump it after ?
I get this message (with connected or disconnected jumper 22) :
![](https://user-images.githubusercontent.com/38383423/120471990-b0e8d080-c3a5-11eb-8d03-0386d2a8142f.jpg)

**Comment by sanni** [comment]
To convert ExLoRom repro PCB to LoRom PCB you have to desolder the A22 jumper and then connect A22(flashrom) to GND with a wire. Then flash as "LoRom (P0)".

![IMG_0151](https://user-images.githubusercontent.com/19169494/121346320-434d1f00-c926-11eb-8d6f-49b7b85a0dde.jpeg)

![IMG_0148](https://user-images.githubusercontent.com/19169494/121346306-40522e80-c926-11eb-9151-408c77472200.JPEG)

If you still get the "Rom header corrupt" message then the rom you have flashed has missing information in the rom header. This can happen with bad rom hacks, betas or test roms. You can still manually dump it by selecting 4MB LoRom in the next menu.

I have also added ExHiRom and ExLoRom options to the "flash repro" menu. Just select an unmodified 6MB rom file like Tales of Phantasia or Fire Emblem Thracia Translation and flash it to an ExHiRom or ExLoRom repro. Please test if it works.

**Reply by StanMOI** [reply]
> I have also added ExHiRom and ExLoRom options to the "flash repro" menu. Just select an unmodified 6MB rom file like Tales of Phantasia or Fire Emblem Thracia Translation and flash it to an ExHiRom or ExLoRom repro. Please test if it works.

Hi Sanni, 
Sorry for late answer.

**Here is test for ExHiRom:**
**PCB config :** FLASH MX29LV640, text on PCB "H01-10k"
Jumper 22 soldered (A22, pin 13 of Flash <=> PCB Golden fingers CART 48)
**ROM:** Tales of Phantasia (translated)
**Size ROM :** unmodified 6MB 
**Write:** Works fine (no error or freezing)
**Verifying:** Works fine! (without freezing at the end)
**Recognizing PCB after writing:** Works fine!
**Dump :** Works fine! (Recognizing rom=ok, 6MB dumped file, tested with "bsnes"=ok) Bad date of created file 31/12/2020.
**TV test:** Works fine!

**Here is test for ExLoRom:**
**PCB config :** FLASH MX29LV640, text on PCB "L-02-rst-R10K"
Jumper 22 soldered (A22, pin 13 of Flash <=> PCB Golden fingers CART 48)
**ROM:** Fire Emblem (translated)
**Size ROM :** unmodified 6MB 
**Write:** Works fine! (no error or freezing)
**Verifying:** Works fine! (without freezing at the end)
**Recognizing PCB after writing:** Works fine!
**Dump :** Works fine! (Recognizing rom=ok, **8MB dumped file and not 6MB**, tested with "bsnes"=ok).  Bad date of created file 31/12/2020.
**TV test:** Works fine!

<img width="510" alt="@stanmoi25 SNES Repro Flash EPROM pinouts copie" src="https://user-images.githubusercontent.com/38383423/123636717-7b58cb00-d81d-11eb-8367-efc4d9b10a8c.png">

**Comment by StanMOI** [comment]
Let me know if you need more tests

**Comment by StanMOI** [comment]
Little update of my pictures with some notes on it for people that it will be useful
<img width="510" alt="SNES Repro 64Mbit Ex-LoROM and Ex-HiROM (256SRAM and 64SRAM) @stanmoi25" src="https://user-images.githubusercontent.com/38383423/137630644-77621e93-cc4a-4cea-a486-4ead02409cc5.png">
![SNES Repro 64Mbit Ex-LoROM (256SRAM)](https://user-images.githubusercontent.com/38383423/137630647-faf73d35-bfca-4f80-a8b3-64a09a68db72.jpeg)
